Target-MVD-Mathias/ViewControllers/SignInViewController.swift

```diff
@@ -10,15 +10,126 @@ import UIKit
+  }
+  
+  // MARK: Actions
   @IBAction func tapOnSignInButton(_ sender: Any) {
-    view.showSpinner(message: "View spinner")
-    UserAPI.login("toptier@mail.com", password: "123456789", success: { (responseObject) in
+    if errorsInForm() {
+      signInError = false
+      showErrorsInForm()
+    } else {
+      signIn()
+    }
+  }
+  
+  // MARK: Private functions\
+  private func errorsInForm() -> Bool {
+    return emailIsInvalid() || passwordIsInvalid()
+  }
+  
+  private func showErrorsInForm() {
+    toggleErrorInForm(error: emailIsInvalid(), textField: emailTextField, errorLabel: emailErrorLabel)
+    toggleErrorInForm(error: passwordIsInvalid(), textField: passwordTextField, errorLabel: signInErrorLabel, errorMessage: passwordErrorMessage)
+  }
+  
+  private func signIn() {
+    showSpinner()
+    resetErrors()
+    
+    let email = emailTextField.text!
+    let password = passwordTextField.text!
+    
+    UserAPI.login(email, password: password, success: { (_) in
       self.hideSpinner()
-      UIApplication.shared.keyWindow?.rootViewController = self.storyboard?.instantiateViewController(withIdentifier: "HomeViewController")
-    }) { (error) in
+      UIApplication.shared.keyWindow?.rootViewController = UIStoryboard.instantiateViewController(HomeViewController.self, storyboardIdentifier: "Onboarding")
+    }) { (_) in
       self.hideSpinner()
-      self.showMessageError(title: "Error", errorMessage: error.localizedDescription)
-      print(error)
+      self.showSignInError()
+      self.signInError = true
+    }
+  }
+  
+  private func showSignInError() {
+    showErrorInForm(textField: emailTextField, errorLabel: nil)
+    showErrorInForm(textField: passwordTextField, errorLabel: signInErrorLabel, errorMessage: signInErrorMessage)
+  }
+  
+  private func resetErrors() {
+    hideErrorInForm(textField: emailTextField, errorLabel: emailErrorLabel)
+    hideErrorInForm(textField: passwordTextField, errorLabel: signInErrorLabel)
+    
+    signInError = false
+  }
+  
+  fileprivate func showErrorInForm(textField: UITextField, errorLabel: UILabel?, errorMessage: String = "") {
+    textField.addBorder(color: .tomato, weight: 1.5)
+    
+    if !errorMessage.isEmpty {
+      errorLabel?.text = errorMessage
+    }
+    
+    errorLabel?.isHidden = false
+  }
+  
+  fileprivate func hideErrorInForm(textField: UITextField, errorLabel: UILabel) {
+    textField.addBorder(color: .black, weight: 1)
+    errorLabel.isHidden = true
+  }
+  
+  fileprivate func emailIsInvalid() -> Bool {
+    let email = emailTextField.text!
+    return !email.isEmailFormatted()
+  }
+  
+  fileprivate func passwordIsInvalid() -> Bool {
+    let password = passwordTextField.text!
+    return password.length() < minPasswordLength
+  }
+  
+}
+
+extension SignInViewController: UITextFieldDelegate {
+  
+  func textFieldDidEndEditing(_ textField: UITextField) {
+    if !signInError {
+      switch textField {
+      case emailTextField:
```

Comment:
In Swift breaks are not necessary.

> In contrast with switch statements in C and Objective-C, switch statements in Swift do not fall through the bottom of each case and into the next one by default. Instead, the entire switch statement finishes its execution as soon as the first matching switch case is completed, without requiring an explicit break statement. This makes the switch statement safer and easier to use than the one in C and avoids executing more than one switch case by mistake.

https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/ControlFlow.html